### ISSUE TITLE
Hot update blows up when adding/removing hooks in React file

### ISSUE THREAD

**Issue by rockcs1992** [pmmmwh/react-refresh-webpack-plugin#514]
Hi,

I'm working on a legacy project and trying to hack modern React dev  tool chain into the environment. I'm using webpack-hot-middleware and following the setup in [this example](https://github.com/pmmmwh/react-refresh-webpack-plugin/tree/main/examples/webpack-hot-middleware).
The basic HMR works well and I could see my changes reflected in the browser. 

![image](https://user-images.githubusercontent.com/16792420/134574060-4138a56f-860d-4071-9835-5cc2cc4ff03a.png)

And my test React file looks like this:

```
const HelloWorld = () => {
    const [count, setCount] = React.useState(0);
    return (
        <div>
            <button onClick={() => setCount(count+1)} >plus</button>
            <div>{count}</div>
        </div>
    );
};

export default HelloWorld;
```

However hot update blows up when I add a hook to the react file:
```
const HelloWorld = () => {
    const [count, setCount] = React.useState(0);
    const [text, _] = React.useState("");
    return (
        <div>
            <button onClick={() => setCount(count+1)} >plus</button>
            <div>{count}</div>
            <div>{text}</div>
        </div>
    );
};
```
 If I make the change and save then it will pop up this hooks error:
![image](https://user-images.githubusercontent.com/16792420/134574326-42d5b831-0a05-40e8-826d-5bc3e3636a5b.png)

I'm clear of the rule of hooks and why this error would pop up in normal development. However I don't understand why it would show up for HMR in this case. and I don't see the error when in your example or in any other HMR repos.

It would be hard for me to duplicate my environment because of the complexity and I do apologize for that. But here is what I could tell how it's different from the other runtime

1. React package is marked as "external" in webpack config and loaded during runtime
2. The react part runs in an Iframe.

Now that I have no clue of why I'm getting the hooks error. Do you by any chance have any clue why this is happening?

Thanks!

**Comment by gaearon** [comment]
Can you try to make a minimal example with an iframe and external React? I'd expect this to be sufficient to repro.

**Comment by pmmmwh** [comment]
Are you sure you have ran the Babel plugin in your project? Have you looked at the guide for externalising React?

**Comment by rockcs1992** [comment]
@gaearon @pmmmwh

**Comment by pmmmwh** [comment]
I've been looking at this but do not have any progress yet. I believe it is likely still related to injection order and Babel processing being incorrect.

**Comment by julienv3** [comment]
Hey everyone,

Tracked this issue down, it's caused by signatures being undefined due to a bug that turns out to have been fixed by @pmmmwh as of version 0.5.1; see change at https://github.com/pmmmwh/react-refresh-webpack-plugin/commit/61985e15f20d8d1c0cea1489807c7078c39290a9.

For @rockcs1992 and @nileshp87, one of the factors deciding if a component can be updated while preserving state is a webpack-plugin-calculated signature which is based on a few factors including `React.useState` calls.
Since at runtime `__webpack_require__.$Refresh$.signature` still pointed to a stub even after requiring `"react-refresh/runtime"` prior to 0.5.1 the signature comparison call was returning early as it had undefined values.

@pmmmwh you may want to bump up your version from 0.5.0-beta.8 to 0.5.1 over at
https://codesandbox.io/s/react-refresh-externals-14fpn?file=/package.json:246-259

**Comment by pmmmwh** [comment]
@julienv3 thanks for digging! I can confirm by bumping to 0.5.1 the error is resolved, on both Webpack 4/5 and React Refresh 0.10.0 or 0.11.0.

**Comment by pmmmwh** [comment]
I'll close this for now but if updating still yields any issues, feel free to re-open.
